stan/math/rev/functor/ad_stack_matrix.hpp

```diff
@@ -0,0 +1,111 @@
+#ifndef STAN_MATH_REV_FUNCTOR_AD_STACK_MATRIX_HPP
+#define STAN_MATH_REV_FUNCTOR_AD_STACK_MATRIX_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/core.hpp>
+
+namespace stan {
+namespace math {
+
+template <typename MatrixType>
+class AD_stack_matrix : public Eigen::Map<MatrixType> {
+ public:
+  using Scalar = value_type_t<MatrixType>;
```

Comment:
`scalar_type_t`?

Reply:
Does not matter as we never nest containers in Eigen matrices. If we did `value_type_t` would be the correct choice.

Reply:
Can it be `value_type` or something like that on the left then?

Reply:
Nope. This is also used by Eigen when this type is used in expressions. Every Eigen type has a Scalar typedef.

Reply:
Then this should be `scalar_type_t` then. I don't think there are plans to nest containers in Eigen matrices.

Reply:
Why? `value_type_t` is both simpler and the one that will still work if we decide to extend this. I see no reason to use `scalar_type_t`.

Reply:
> Why?

I just want scalars to be scalars to be scalars :(, but okay, we'll go with this.